doc/design/parallel_executor.md

```diff
@@ -0,0 +1,53 @@
+# ParallelExecutor Design Doc
+
+## Introduction
+
+We introduce `ParallelExecutor`, as an alternative solution to `ParallelDo`, to run multi-GPU
+training in PaddlePaddle Fluid. Essentially, it is a python wrapper of a list of `Executor`s
+and `Scopes`s. In addition, it maintains inter-device variables such as NCCL communicator.
+
+To train a neural network multi GPUs, a user only need to make the following modifications in
+the code
+1. specify `append_all_reduce` flag in the optimizer
+1. use `ParallelExecutor` to run the `programDesc`.
+
+```python
+cost = your_neural_network()
+
+opt = fluid.optimizer.SGDOptimizer(..., append_all_reduce=True)
```

Comment:
what if append_all_reduce=True is not set, will it use other way to send gradients to each other?

Reply:
Currently, it's not handled.